### PR TITLE
add blobKeepAliveTimeout, queueKeepAliveTimeout, tableKeepAliveTimeout 

### LINKED ISSUE #2053: unexptected TCP RST from azurite side
### Which service(blob, file, queue, table) does this issue concern?
blob
### Which version of the Azurite was used?
latest docker image
### Where do you get Azurite? (npm, DockerHub, NuGet, Visual Studio Code Extension)
I created reproducible docker-compose
https://gist.github.com/Slach/043c85f759d4422a75fceab6dccb1f8d

### What problem was encountered?
TCP RST after make client requests
image
image

### Steps to reproduce the issue?
```
git clone https://gist.github.com/Slach/043c85f759d4422a75fceab6dccb1f8d
./reproduce.sh
```

### Have you found a mitigation/solution?
No

also look details https://github.com/ClickHouse/ClickHouse/issues/52088

### PR DESCRIPTION
**Pull request by Slach**
second try after https://github.com/Azure/Azurite/pull/2443

fix #2053 and workaround ClickHouse/ClickHouse#60447

### CODE REVIEW

**Comment by blueww** [comment]
@Slach 

Would you please share more background of this change?
1. Does it target for a temp fix of CPP SDK issue
2. Does product Azure server has same behavior after the change?

If it target a temp fix, or will make Azurite behavior not aligned with product Azure, as I have indicated in the before PR https://github.com/Azure/Azurite/pull/2443#issuecomment-2268057651, we will take the change to public Azurite, you can build your own private Azurite as a workaround.

**Comment by Slach** [comment]
1. This change allows properly using `azurite` in e2e test scenarios with old version of azure cpp-sdk
current version of cpp sdk still contains issue when client dont respect `Keep-alive` header from server

2. Yes, by default `azurite` will have the same behavior as with the previous version, when send `Keep-alive: timeout=5`

**Comment by Slach** [comment]
You can't upgrade azure cpp sdk version in already released products 

too much work for backports even when new cpp sdk will release

**Comment by blueww** [comment]
> 1. This change allows properly using `azurite` in e2e test scenarios with old version of azure cpp-sdk
>    current version of cpp sdk still contains issue when client dont respect `Keep-alive` header from server
> 2. Yes, by default `azurite` will have the same behavior as with the previous version, when send `Keep-alive: timeout=5`

for #2, if user set none default value, will Azurite behavior different from product Azure? 

If this is only for a temp SDK issue fix, Azurite won't take it. Especially when the change could make Azurite behavior different from product Azure.

**Comment by Slach** [comment]
> If this is only for a temp SDK issue fix

this is not temp SDK issue fix.

you can't guarantee same behavior from other SDK
issue become to your product from standard nodejs

**Comment by blueww** [comment]
> > If this is only for a temp SDK issue fix
> 
> this is not temp SDK issue fix.
> 
> You can't guarantee same behavior from other SDKs
> 
> issue become to your product from standard nodejs [nodejs/node#34560](https://github.com/nodejs/node/issues/34560)

Does the client SDK with this issue works with product Azure Server?
And  if user set none default value, will Azurite behavior different from product Azure?

**Comment by Slach** [comment]
> for second point, if user set none default value, will Azurite behavior different from product Azure?

when you pass --blobKeepAliveTimeout=0 or --blobKeepAliveTimeout=""
then this parameter will just ignore and nodejs httpServer will initialize with default

Actually, `Azurite` is different with Azure in this part, right now, without this PR

`Azurite` already send `Keep-alive: timeout=5` header in server response
but `Azure` doesn't

check how exactly nodejs handle Keep-Alive 

`Azure` example request+response
```
   PUT https://xxxbackups.blob.core.windows.net/xxxbackups/altinity-cloud-managed-clickhouse/backup-test/full_backup_8596655552399465575/shadow/test_partitions_TestIntegrationAzure/t2/default_435ef8e8df48e77cefdcc330dc8b8d8c_3_3_0.tar?blockid=lPNYkFCYSziFsOQF4bwpJQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%3D%3D&comp=block&timeout=14401
   Authorization: REDACTED
   Content-Length: [12288]
   User-Agent: [Azure-Storage/0.15 (go1.22.6; linux)]
   X-Ms-Client-Request-Id: [6cfaab65-8c46-4f48-68b5-a992ec7524ee]
   X-Ms-Version: [2020-10-02]
   x-ms-date: [Wed, 28 Aug 2024 12:31:05 GMT]
   --------------------------------------------------------------------------------
   RESPONSE Status: 201 Created
   Content-Length: [0]
   Date: [Wed, 28 Aug 2024 12:29:30 GMT]
   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]
   X-Ms-Client-Request-Id: [6cfaab65-8c46-4f48-68b5-a992ec7524ee]
   X-Ms-Content-Crc64: [ppWapnj/W7U=]
   X-Ms-Request-Id: [9b01ff55-d01e-006f-0445-f946cc000000]
   X-Ms-Request-Server-Encrypted: [true]
   X-Ms-Version: [2020-10-02]
```   

`Azurite` for the same code
```
PUT http://devstoreaccount1.blob.azure:10000/container1/backup/full_backup_2226074798298314988/shadow/test_partitions_TestIntegrationAzure/t1/default_0%252D20220102_2_2_0.tar?blockid=sCJz0hyuTlWyUK7qdIfDHgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA%3D%3D&comp=block&timeout=14401
   Authorization: REDACTED
   Content-Length: [12288]
   User-Agent: [Azure-Storage/0.15 (go1.22.6; linux)]
   X-Ms-Client-Request-Id: [161470f4-dc6f-41e4-6ea7-6f2e39b7ad9d]
   X-Ms-Version: [2020-10-02]
   x-ms-date: [Wed, 28 Aug 2024 12:45:10 GMT]
   --------------------------------------------------------------------------------
   RESPONSE Status: 201 Created
   Connection: [keep-alive]
   Content-Length: [0]
   Date: [Wed, 28 Aug 2024 12:45:10 GMT]
   Keep-Alive: [timeout=5]
   Server: [Azurite-Blob/3.32.0]
   X-Ms-Client-Request-Id: [161470f4-dc6f-41e4-6ea7-6f2e39b7ad9d]
   X-Ms-Request-Id: [bd712af4-70f8-4c94-aecd-af01199f6399]
   X-Ms-Request-Server-Encrypted: [true]
   X-Ms-Version: [2024-11-04]
```

**Comment by blueww** [comment]
@Slach 

Thanks for the details!

I see Azurite will return 2 more headers than Azure server :
```
Connection: keep-alive
Keep-Alive: timeout=5
```

In this case, it's better we make the default Azurite behavior aligned with Azure server as not return the 2 headers. Would you please help to update the PR for this?

We will review/test the PR and update you later.

File: src/queue/QueueServer.ts [hunk 1]
```diff
@@ -70,6 +70,10 @@ export default class QueueServer extends ServerBase {
         httpServer = http.createServer();
     }
 
+    if (configuration.keepAliveTimeout > 0) {
+      httpServer.keepAliveTimeout = configuration.keepAliveTimeout
```
Comment (blueww) [inline src/queue/QueueServer.ts:74]: Why one update Queue, but not update blob/table?

**Comment by Slach** [comment]
> In this case, it's better we make the default Azurite behavior aligned with Azure server as not return the 2 headers. Would you please help to update the PR for this?

could you read details in https://github.com/nodejs/node/issues/34560 ?

this is nodejs standard library behavior for any nodejs application which use `httpServer`, unfortunately i not so familiar with nodejs to make PR to [no](https://github.com/nodejs/node)

need time to think how to do it, current PR just allow us avoid conflicts with cpp-sdk

File: src/queue/QueueServer.ts (hunk 1 above)
Comment (Slach) [reply src/queue/QueueServer.ts:74]: blob and table updated in different place
look
https://github.com/Azure/Azurite/pull/2454/files#diff-de9f719b61be179ac28afd68eb6902e9c4011b46745f114249ae6441a15ab086R50

your code base for queue and table/blob just inconsistent in this part

File: README.md [hunk 2]
```diff
@@ -186,10 +186,13 @@ Following extension configurations are supported:
 
 - `azurite.blobHost` Blob service listening endpoint, by default 127.0.0.1
 - `azurite.blobPort` Blob service listening port, by default 10000
+- `azurite.blobKeepAliveTimeout` Blob service keep alive timeout, by default 5
```
Comment (blueww) [inline README.md:189]: The default should be 5 or 5000?
And what's the valid value range? (look 0 won't take effect)

File: README.md [hunk 3]
```diff
@@ -186,10 +186,13 @@ Following extension configurations are supported:
...
 - `azurite.queueHost` Queue service listening endpoint, by default 127.0.0.1
 - `azurite.queuePort` Queue service listening port, by default 10001
+- `azurite.queueKeepAliveTimeout` Queue service keep alive timeout, by default 5
 - `azurite.tableHost` Table service listening endpoint, by default 127.0.0.1
 - `azurite.tablePort` Table service listening port, by default 10002
+- `azurite.tableKeepAliveTimeout` Queue service keep alive timeout, by default 5
```
Comment (blueww) [inline README.md:195]: If the 3 parameters also works for docker, pleas also add them to docker parameter list in this file.

File: src/queue/QueueServer.ts (hunk 1 above)
Comment (blueww) [reply src/queue/QueueServer.ts:74]: Thanks for the answer!

File: README.md (hunk 2 above)
Comment (Slach) [reply README.md:189]: 5000 i will fix it ASAP

File: README.md (hunk 3 above)
Comment (Slach) [reply README.md:195]: ok. will do

**Comment by blueww** [comment]
> > In this case, it's better we make the default Azurite behavior aligned with Azure server as not return the 2 headers. Would you please help to update the PR for this?
> 
> could you read details in [nodejs/node#34560](https://github.com/nodejs/node/issues/34560) ?
> 
> this is nodejs standard library behavior for any nodejs application which use `httpServer`, unfortunately i not so familiar with nodejs to make PR to [nodejs](https://github.com/nodejs/node)
> 
> need time to think how to do it, current PR just allow us avoid conflicts with cpp-sdk

Hope we could make the default behavior works like Azure server which won't return the headers.
If really can't find a way to get so, keep the default as current behavior should be the choice.

**Comment by Slach** [comment]
> If really can't find a way to get so, keep the default as current behavior should be the choice.

This is my  working plan

File: README.md (hunk 2 above)
Comment (blueww) [reply README.md:189]: Please also update the doc, to include the default value and valid value range. It's better to also include the impact of set the paramater.

File: README.md (hunk 3 above)
Comment (Slach) [reply README.md]: fixed in https://github.com/Azure/Azurite/pull/2454/commits/5a147d5b8ddf187fcf392908e4811a4a08088cf0

File: README.md (hunk 2 above)
Comment (Slach) [reply README.md]: fixed in https://github.com/Azure/Azurite/pull/2454/commits/c66120e9c6a48c532295829b97a1ae4769af39f1

**Comment by Slach** [comment]
@blueww could you trigger github actions?

File: README.md [hunk 4]
```diff
@@ -186,10 +186,13 @@ Following extension configurations are supported:
 
 - `azurite.blobHost` Blob service listening endpoint, by default 127.0.0.1
 - `azurite.blobPort` Blob service listening port, by default 10000
+- `azurite.blobKeepAliveTimeout` Blob service keep alive timeout, by default 5000
```
Comment (blueww) [inline README.md:189]: Could more description add here? Like 5000 means 5 second. And the impact by adding this parameter.

File: tests/BlobTestServerFactory.ts [hunk 5]
```diff
@@ -4,6 +4,7 @@ import SqlBlobConfiguration from "../src/blob/SqlBlobConfiguration";
 import SqlBlobServer from "../src/blob/SqlBlobServer";
 import { StoreDestinationArray } from "../src/common/persistence/IExtentStore";
 import { DEFAULT_SQL_OPTIONS } from "../src/common/utils/constants";
+import { DEFAULT_BLOB_KEEP_ALIVE_TIMEOUT } from "../src/blob/utils/constants";
```
Comment (blueww) [inline tests/BlobTestServerFactory.ts:7]: Could we add test case for B/Q/T service to check adding this parameter works.

**Comment by blueww** [comment]
@Slach 
I believe the PR is ready to review after you update it.
Others are good to me.  Just add 2 comments for the parameter description and test cases.

File: README.mcr.md [hunk 6]
```diff
@@ -80,6 +80,12 @@ Above command will try to start Azurite image with configurations:
 
 Please refer to this [document](https://github.com/Azure/Azurite/blob/master/README.md) for **More supported parameters** like HTTPS or OAuth.
 
+**Customize HTTP Keep-Alive behavior**
+
+```bash
+docker run azurite --blobHost 0.0.0.0 --blobKeepAliveTimeout 300 --queueHost 0.0.0.0 --queueKeepAliveTimeout 300  --tableHost 0.0.0.0 --tableKeepAliveTimeout 300
```
Comment (blueww) [inline README.mcr.md:86]: for the "--blobKeepAliveTimeout 300" here, do you mean 300s?
If so, per current implementation, it should be 300000.

And please also update the description to say Customize HTTP Keep-Alive to how long time.

**Comment by blueww** [comment]
@Slach

Could you please look at my comments above, and see if can fix them.
Feel free to let me know if you have different opinion.

**Comment by Slach** [comment]
sorry i'm on vacation will fix when return

File: tests/BlobTestServerFactory.ts (hunk 5 above)
Comment (Slach) [reply tests/BlobTestServerFactory.ts:7]: Sorry for late response, could you suggest how to make this TestCase? How to check response headers to Keepalive-Timeout header exists and contains properly value?

File: README.mcr.md (hunk 6 above)
Comment (Slach) [reply README.mcr.md:87]: i changed it to seconds and switch default constant values to seconds

File: README.md (hunk 4 above)
Comment (Slach) [reply README.md]: I figure out and switched from ms to seconds, look https://github.com/Azure/Azurite/pull/2454/commits/11704a31baf6091e08bdbd09270bd42afae8e60f

**Comment by Slach** [comment]
Hi @blueww, i prepared some changes need your advice about tests
https://github.com/Azure/Azurite/pull/2454#discussion_r1798951345

current test code base is too complicated for my understanding
i tested changes internally manually and it works

File: tests/BlobTestServerFactory.ts (hunk 5 above)
Comment (blueww) [reply tests/BlobTestServerFactory.ts:7]: @EmmaZhu 

Could you please share how to get a header in raw server responds with JS SDK?

Comment (Slach) [reply tests/BlobTestServerFactory.ts:7]: @EmmaZhu any news from your side?

Comment (EmmaZhu) [reply tests/BlobTestServerFactory.ts:7]: Hi @Slach ,

You can use code like following to get keep-alive header:

```
    const properties = await newBlobClient.getProperties();
    const keepAliveHeader = properties._response.headers.get("keep-alive");
```

File: src/common/VSCServerManagerQueue.ts [hunk 7]
```diff
@@ -77,6 +77,7 @@ export default class VSCServerManagerBlob extends VSCServerManagerBase {
     const config = new QueueConfiguration(
       env.queueHost(),
       env.queuePort(),
+      env.blobKeepAliveTimeout(),
```
Comment (blueww) [inline src/common/VSCServerManagerQueue.ts:80]: should be `env.queueKeepAliveTimeout(),`

Comment (Slach) [reply src/common/VSCServerManagerQueue.ts:80]: sorry for late response, fixed but why `VSCServerManagerBlob` class name in this file?

File: tests/BlobTestServerFactory.ts (hunk 5 above)
Comment (Slach) [reply tests/BlobTestServerFactory.ts:7]: @EmmaZhu
sorry to late reply

this approach doesn't work for table service

following code fragment doesn't work
```
    const tableServiceClient = new TableServiceClient(baseURL, sharedKeyCredential);
    const properties = await tableServiceClient.getProperties();
    const keepAliveHeader = properties._response.headers.get("keep-alive");

```

`TS2339: Property '_response' does not exist on type 'GetPropertiesResponse'.`

How can we get keeper-alive header from server response for Table?

Comment (Slach) [reply tests/BlobTestServerFactory.ts:7]: @blueww I added test (and it passed) for blob and queue
look https://github.com/Azure/Azurite/pull/2454/commits/ee051935787ca9304dcdbdcb52a3a26b11916c1e
but need suggestion how to get headers for `Table`

could you trigger CI/CD pipeline on your side?

Comment (Slach) [reply tests/BlobTestServerFactory.ts:7]: ok. added `tableKeepAliveTimeout.test.js` https://github.com/Azure/Azurite/pull/2454/commits/8bdd4b5c1ad2b5db8994832f72a2e02255248e18

**Comment by Slach** [comment]
@blueww could you trigger CI/CD? Hope PR ready for merge

File: README.md [hunk 8]
```diff
@@ -246,14 +249,20 @@ Above command will try to start Azurite image with configurations:
 
 `--blobHost 0.0.0.0` defines blob service listening endpoint to accept requests from host machine.
 
+`--blobKeepAliveTimeout 5` blob service keep alive timeout in seconds
```
Comment (blueww) [inline README.md:252]: The section is to explain the parameter used in sample [in ling #236](https://github.com/Azure/Azurite/blob/242e14087333c9c3cf9f6ca3e2328d8f25981c9c/README.md#customize-all-azurite-v3-supported-parameters-for-docker-image)

However the sample is not updated with the new parameters, so add the description here looks odd. 
You might can update the sample with the new parameters?

Comment (Slach) [reply README.md:252]: fixed in https://github.com/Azure/Azurite/pull/2454/commits/d63a8902b27dde5ff309411815444a33784a107d

**Comment by Slach** [comment]
@XiaoningLiu, @EmmaZhux any news from your side?

**Comment by Slach** [comment]
@XiaoningLiu, @EmmaZhu could you merge this PR?

**Review by XiaoningLiu — APPROVED** [review]

**Pull request merged**